src/mlmc/quantity_concept.py

```diff
@@ -1,115 +1,430 @@
+
+    return qtype
+
+
+def _create_quantity(values, base_qtype, quantity_storage):
+    """
+    Convert flat, bool or array (list) to Quantity
+    :param values: flat, bool, array (list)
+    :param base_qtype: QType
+    :return: Quantity
+    """
+    if isinstance(values, (int, float, bool)):
+        quantity = QuantityConst(quantity_type=base_qtype, values=values, input_quantities=[quantity_storage])
+    elif isinstance(values, (list, np.ndarray)):
+        values = np.array(values)
+        qtype = ArrayType(shape=values.shape[0], qtype=base_qtype)
+        quantity = QuantityConst(quantity_type=qtype, values=values, input_quantities=[quantity_storage])
+    else:
+        raise ValueError("Values {} are not flat, bool or array (list)".format(values))
+    return quantity
+
+
+def _method(ufunc, method, *args, **kwargs):
+    """
+    Process input parameters to perform numpy ufunc.
+    Get base QType of passed quantities, QuantitiStorage instance, ...
+    Determine the resulting QType from the first few samples
+    :param ufunc: ufunc object that was called
+    :param method: string, indicating which Ufunc method was called
+    :param args: tuple of the input arguments to the ufunc
+    :param kwargs: dictionary containing the optional input arguments of the ufunc
+    :return: Quantity
+    """
+    def _demo_method(*input_quantities_chunks):
+        if len(input_quantities_chunks) == 1:
+            return getattr(ufunc, method)(input_quantities_chunks[0], **kwargs)
+        else:
+            return getattr(ufunc, method)(*input_quantities_chunks, **kwargs)
+
+    base_qtype, quantity_storage = _get_quantity_info(args, get_quantity_storage=True)
+
+    quantities = []
+    for arg in args:
+        if not isinstance(arg, Quantity):
+            quantities.append(_create_quantity(arg, base_qtype=base_qtype, quantity_storage=quantity_storage))
+        else:
+            quantities.append(arg)
+
+    # Check if all quantities come from same storage
+    if not all(np.allclose(q.storage_id(), quantities[0].storage_id()) for q in quantities[1:]):
+        raise Exception("Not all input quantities come from the same quantity storage")
+
+    result_qtype = _determine_qtype(quantities, _demo_method)
+    return Quantity(quantity_type=result_qtype, input_quantities=list(quantities), operation=_demo_method)
+
+
+def make_root_quantity(storage: SampleStorage, q_specs: List[QuantitySpec]):
+    """
+    Create a root quantity that has QuantityStorage as the input quantity,
+    QuantityStorage is the only class that directly accesses the stored data.
+    Quantity type is created based on the q_spec parameter
+    :param storage: SampleStorage
+    :param q_specs: same as result format in simulation class
+    :return: QuantityStorage
+    """
+    # Set chunk size as the case may be
+    if storage.chunk_size is None:
+        storage.chunk_size = 512000  # bytes in decimal
```

Comment:
- Make this a named constant. 
- Setting the chunk size to the storage seems inapropriate, possibly we can introduce a small object
ChunkSpec, that could be passed through the samples methods down to the storage and then the storage returns the appropriate chunk. Then the chunk size can be (optional) parameter of the `estimate_mean` which is better since then we can possibly set the chunk_size dinamically according to the free memory or cache size ...
Moreover this can be used to get try chunk for determining the result type of the ufunc.

Reply:
separate issue created